File: apps/ios/TVM/TVMCatalog.swift
import Foundation

struct CatalogBundle {
    var moviesTop: [MediaItem]
    var seriesTop: [MediaItem]
    var moviesRated: [MediaItem]
    var seriesRated: [MediaItem]
    var recentTv: [MediaItem]
    var newFilms: [MediaItem]
    var catalog: [MediaItem]
}

struct TitleMeta {
    var item: MediaItem
    var children: [MediaItem]
}

final class TVMCatalog {
    private let store: TVMStore
    private let session: URLSession
    private var memory: (at: Date, data: CatalogBundle)?
    private var genreMem: [String: (at: Date, items: [MediaItem])] = [:]
    private var metaMem: [String: (at: Date, data: TitleMeta)] = [:]
    private let fallback: [MediaItem]
    private let lock = NSLock()

    init(store: TVMStore, session: URLSession) {
        self.store = store
        self.session = session
        if let url = Bundle.main.url(forResource: "FallbackCatalog", withExtension: "json"),
           let data = try? Data(contentsOf: url),
           let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let items = object["items"] as? [Any] {
            fallback = items.compactMap(MediaItem.parse)
        } else {
            fallback = []
        }
    }

    func fallbackItems() -> [MediaItem] { fallback }

    func clear() {
        lock.lock()
        memory = nil
        genreMem.removeAll()
        metaMem.removeAll()
        lock.unlock()
        store.clearCacheFiles()
    }

    func bundle() async -> CatalogBundle {
        lock.lock()
        if let memory, Date().timeIntervalSince(memory.at) < 12 * 60 {
            let data = memory.data
            lock.unlock()
            return data
        }
        lock.unlock()
        do {
            async let moviesTopTask = fetchCatalog("/catalog/movie/top.json", kind: "movie")
            async let seriesTopTask = fetchCatalog("/catalog/series/top.json", kind: "series")
            async let moviesRatedTask = fetchCatalog("/catalog/movie/imdbRating.json", kind: "movie")
            async let seriesRatedTask = fetchCatalog("/catalog/series/imdbRating.json", kind: "series")
            async let recentTvTask = fetchCatalog("/catalog/series/last-videos.json", kind: "series")
            async let moviesSkipTask = fetchCatalog("/catalog/movie/top/skip=100.json", kind: "movie")
            let moviesTop = await moviesTopTask
            let seriesTop = await seriesTopTask
            let moviesRated = await moviesRatedTask
            let seriesRated = await seriesRatedTask
            let recentTv = await recentTvTask
            let moviesSkip = await moviesSkipTask
            let movies = dedupe(moviesTop + moviesSkip)
            let data = CatalogBundle(
                moviesTop: moviesTop,
                seriesTop: seriesTop,
                moviesRated: moviesRated,
                seriesRated: seriesRated,
                recentTv: recentTv,
                newFilms: movies.filter { $0.kind == "movie" && ($0.year ?? 0) >= Calendar.current.component(.year, from: Date()) - 1 },
                catalog: dedupe(movies + seriesTop + moviesRated + seriesRated + recentTv)
            )
            if !data.catalog.isEmpty {
                lock.lock()
                memory = (Date(), data)
                lock.unlock()
                store.writeJSON("catalog-cache.json", encode(data))
                return data
            }
        } catch {
            // Use disk or bundled titles so Home still renders.
        }
        if let disk = readDisk() { return disk }
        return fallbackBundle()
    }

    func search(_ query: String) async -> [MediaItem] {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines)
        if needle.count < 2 { return [] }
        let encoded = needle.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? needle
        async let movies = fetchCatalog("/catalog/movie/top/search=\(encoded).json", kind: "movie")
        async let series = fetchCatalog("/catalog/series/top/search=\(encoded).json", kind: "series")
        let live = dedupe(await movies + series)
        if !live.isEmpty { return live }
        let lower = needle.lowercased()
        return fallback.filter {
            $0.title.lowercased().contains(lower) || ($0.showTitle ?? "").lowercased().contains(lower)
        }
    }

    func genre(kind: String, name: String) async -> [MediaItem] {
        let key = "\(kind):\(name)"
        lock.lock()
        if let hit = genreMem[key], Date().timeIntervalSince(hit.at) < 20 * 60 {
            let items = hit.items
            lock.unlock()
            return items
        }
        lock.unlock()
        let encoded = name.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? name
        let items = await fetchCatalog("/catalog/\(kind)/top/genre=\(encoded).json", kind: kind)
        lock.lock()
        genreMem[key] = (Date(), items)
        lock.unlock()
        return items
    }

    func meta(_ id: String) async -> TitleMeta? {
        guard let imdb = TVMTitle.extractImdb(id) else {
            return fallback.first { $0.id == id }.map { TitleMeta(item: $0, children: []) }
        }
        lock.lock()
        if let hit = metaMem[imdb], Date().timeIntervalSince(hit.at) < 30 * 60 {
            let data = hit.data
            lock.unlock()
            return data
        }
        lock.unlock()
        async let movie = readKind("movie", imdb: imdb)
        async let series = readKind("series", imdb: imdb)
        let picked = (await series) ?? (await movie)
        if let picked {
            lock.lock()
            metaMem[imdb] = (Date(), picked)
            lock.unlock()
        }
        return picked
    }

    private func readKind(_ kind: String, imdb: String) async -> TitleMeta? {
        guard let url = URL(string: "https://v3-cinemeta.strem.io/meta/\(kind)/\(imdb).json") else { return nil }
        do {
            let (data, response) = try await session.data(from: url)
            guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode),
                  let body = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let raw = body["meta"] as? [String: Any],
                  let item = mapMeta(raw, kind: kind) else { return nil }
            let children = kind == "series" ? mapVideos(show: item, videos: raw["videos"]) : []
            return TitleMeta(item: item, children: children)
        } catch {
            return nil
        }
    }

    private func fetchCatalog(_ path: String, kind: String) async -> [MediaItem] {
        guard let url = URL(string: "https://v3-cinemeta.strem.io\(path)") else { return [] }
        var request = URLRequest(url: url)
        request.timeoutInterval = 12
        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode),
                  let body = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let metas = body["metas"] as? [[String: Any]] else { return [] }
            return metas.compactMap { mapMeta($0, kind: kind) }
        } catch {
            return []
        }
    }

    private func mapMeta(_ raw: [String: Any], kind: String) -> MediaItem? {
        let id = String(describing: raw["id"] ?? raw["imdb_id"] ?? "")
        guard id.range(of: #"^tt\d+$"#, options: .regularExpression) != nil else { return nil }
        let title = ((raw["name"] as? String) ?? (raw["title"] as? String) ?? "").trimmingCharacters(in: .whitespaces)
        if title.isEmpty { return nil }
        let genres = (raw["genre"] as? [String]) ?? []
        let poster = (raw["poster"] as? String) ?? ""
        let backdrop = (raw["background"] as? String) ?? (raw["backdrop"] as? String) ?? poster
        let runtime: String?
        if let text = raw["runtime"] as? String { runtime = text }
        else if let number = raw["runtime"] as? Int { runtime = "\(number) min" }
        else { runtime = nil }
        return MediaItem(
            id: id.lowercased(),
            title: title,
            year: parseYear(raw["year"]) ?? parseYear(raw["releaseInfo"]),
            kind: kind,
            synopsis: raw["description"] as? String ?? "",
            poster: poster,
            backdrop: backdrop,
            genres: genres,
            rating: raw["imdbRating"] == nil ? "" : String(describing: raw["imdbRating"]!),
            runtime: runtime,
            playable: true,
            progress: nil,
            filename: nil,
            hue: TVMTitle.hue(for: title),
            mimeType: nil,
            season: nil,
            episode: nil,
            episodeName: nil,
            showTitle: title,
            aired: nil,
            added: nil
        )
    }

    private func mapVideos(show: MediaItem, videos: Any?) -> [MediaItem] {
        guard let list = videos as? [[String: Any]] else { return [] }
        return list.compactMap { video -> MediaItem? in
            let fromId = String(describing: video["id"] ?? "")
            let idParts = fromId.split(separator: ":")
            let season = intValue(video["season"] ?? video["seasonNumber"]) ?? (idParts.count >= 3 ? Int(idParts[idParts.count - 2]) : nil)
            let episode = intValue(video["episode"] ?? video["number"] ?? video["episodeNumber"]) ?? (idParts.last.flatMap { Int($0) })
            guard let season, let episode, season >= 1, episode >= 1 else { return nil }
            let name = (video["title"] as? String) ?? (video["name"] as? String) ?? "Episode \(episode)"
            let thumbnail = (video["thumbnail"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? show.poster
            return MediaItem(
                id: "\(show.id):\(season):\(episode)",
                title: show.title,
                year: show.year,
                kind: "series",
                synopsis: (video["overview"] as? String) ?? (video["description"] as? String) ?? "",
                poster: thumbnail,
                backdrop: show.backdrop,
                genres: show.genres,
                rating: "",
                runtime: nil,
                playable: true,
                progress: nil,
                filename: nil,
                hue: show.hue,
                mimeType: nil,
                season: season,
                episode: episode,
                episodeName: name,
                showTitle: show.title,
                aired: nil,
                added: nil
            )
        }.sorted {
            ($0.season ?? 0, $0.episode ?? 0) < ($1.season ?? 0, $1.episode ?? 0)
        }
    }

    private func parseYear(_ value: Any?) -> Int? {
        if let number = value as? Int, number > 1800 { return number }
        if let text = value as? String, let match = text.range(of: #"\b(19|20)\d{2}\b"#, options: .regularExpression) {
            return Int(text[match])
        }
        return nil
    }

    private func intValue(_ value: Any?) -> Int? {
        if let number = value as? Int { return number }
        if let number = value as? Double { return Int(number) }
        if let text = value as? String { return Int(text) }
        return nil
    }

    private func dedupe(_ items: [MediaItem]) -> [MediaItem] {
        var seen = Set<String>()
        return items.filter { seen.insert($0.id).inserted }
    }

    private func fallbackBundle() -> CatalogBundle {
        let movies = fallback.filter { $0.kind == "movie" }
        let series = fallback.filter { $0.kind == "series" }
        return CatalogBundle(
            moviesTop: Array(movies.prefix(16)),
            seriesTop: Array(series.prefix(16)),
            moviesRated: Array(movies.prefix(16)),
            seriesRated: Array(series.prefix(16)),
            recentTv: Array(series.prefix(16)),
            newFilms: Array(movies.filter { ($0.year ?? 0) >= 2023 }.prefix(16)),
            catalog: fallback
        )
    }

    private func readDisk() -> CatalogBundle? {
        guard let object = store.readJSON("catalog-cache.json") as? [String: Any],
              let movies = object["moviesTop"] as? [Any] else { return nil }
        return CatalogBundle(
            moviesTop: movies.compactMap(MediaItem.parse),
            seriesTop: (object["seriesTop"] as? [Any] ?? []).compactMap(MediaItem.parse),
            moviesRated: (object["moviesRated"] as? [Any] ?? []).compactMap(MediaItem.parse),
            seriesRated: (object["seriesRated"] as? [Any] ?? []).compactMap(MediaItem.parse),
            recentTv: (object["recentTv"] as? [Any] ?? []).compactMap(MediaItem.parse),
            newFilms: (object["newFilms"] as? [Any] ?? []).compactMap(MediaItem.parse),
            catalog: (object["catalog"] as? [Any] ?? []).compactMap(MediaItem.parse)
        )
    }

    private func encode(_ data: CatalogBundle) -> [String: Any] {
        [
            "moviesTop": data.moviesTop.map { $0.json() },
            "seriesTop": data.seriesTop.map { $0.json() },
            "moviesRated": data.moviesRated.map { $0.json() },
            "seriesRated": data.seriesRated.map { $0.json() },
            "recentTv": data.recentTv.map { $0.json() },
            "newFilms": data.newFilms.map { $0.json() },
            "catalog": data.catalog.map { $0.json() },
        ]
    }
}

File: apps/android/app/src/main/java/com/tvm/privateclient/TvmCatalog.kt
package com.tvm.privateclient

import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URI
import java.util.Calendar
import java.util.Locale
import java.util.concurrent.Callable
import java.util.concurrent.Executors

/**
 * Cinemeta browse data, cached in memory, on disk and finally in the bundle.
 *
 * Direct port of apps/ios/TVM/TVMCatalog.swift. The same `apps/ui` interface
 * reads both clients, so the shapes, defaults and fallbacks here must match the
 * Swift exactly.
 */

data class CatalogBundle(
    val moviesTop: List<MediaItem>,
    val seriesTop: List<MediaItem>,
    val moviesRated: List<MediaItem>,
    val seriesRated: List<MediaItem>,
    val recentTv: List<MediaItem>,
    val newFilms: List<MediaItem>,
    val catalog: List<MediaItem>,
)

data class TitleMeta(val item: MediaItem, val children: List<MediaItem>)

/**
 * @param store the same on-disk store the rest of the client uses; `catalog-cache.json` lives there.
 * @param loadBundled reads the bundled FallbackCatalog.json. The caller supplies it so this file
 *   stays free of android.content.Context and testable on the JVM, e.g.
 *   `{ runCatching { context.assets.open("FallbackCatalog.json").bufferedReader().use { it.readText() } }.getOrNull() }`.
 *   It is called once, from this constructor, exactly where the Swift reads the bundle resource.
 */
class TvmCatalog(
    private val store: TvmStore,
    loadBundled: () -> String? = { null },
) {
    private val lock = Any()
    private var memoryAt: Long = 0
    private var memory: CatalogBundle? = null
    private val genreMem: MutableMap<String, Pair<Long, List<MediaItem>>> = mutableMapOf()
    private val metaMem: MutableMap<String, Pair<Long, TitleMeta>> = mutableMapOf()
    private val fallback: List<MediaItem>

    init {
        val text = runCatching { loadBundled() }.getOrNull()
        val items = if (text.isNullOrEmpty()) null else Json.parseObject(text).optJSONArray("items")
        fallback = if (items == null) emptyList() else parseItems(items)
    }

    fun fallbackItems(): List<MediaItem> = fallback

    fun clear() {
        synchronized(lock) {
            memory = null
            memoryAt = 0
            genreMem.clear()
            metaMem.clear()
        }
        store.clearCacheFiles()
    }

    suspend fun bundle(): CatalogBundle {
        synchronized(lock) {
            val cached = memory
            if (cached != null && System.currentTimeMillis() - memoryAt < 12 * 60 * 1000L) return cached
        }
        val fetched: List<List<MediaItem>> = runParallel(
            listOf(
                { fetchCatalog("/catalog/movie/top.json", "movie") },
                { fetchCatalog("/catalog/series/top.json", "series") },
                { fetchCatalog("/catalog/movie/imdbRating.json", "movie") },
                { fetchCatalog("/catalog/series/imdbRating.json", "series") },
                { fetchCatalog("/catalog/series/last-videos.json", "series") },
                { fetchCatalog("/catalog/movie/top/skip=100.json", "movie") },
            ),
            emptyList(),
        )
        val moviesTop = fetched[0]
        val seriesTop = fetched[1]
        val moviesRated = fetched[2]
        val seriesRated = fetched[3]
        val recentTv = fetched[4]
        val moviesSkip = fetched[5]
        val movies = dedupe(moviesTop + moviesSkip)
        val thisYear = Calendar.getInstance().get(Calendar.YEAR)
        val data = CatalogBundle(
            moviesTop = moviesTop,
            seriesTop = seriesTop,
            moviesRated = moviesRated,
            seriesRated = seriesRated,
            recentTv = recentTv,
            newFilms = movies.filter { it.kind == "movie" && (it.year ?: 0) >= thisYear - 1 },
            catalog = dedupe(movies + seriesTop + moviesRated + seriesRated + recentTv),
        )
        if (data.catalog.isNotEmpty()) {
            synchronized(lock) {
                memory = data
                memoryAt = System.currentTimeMillis()
            }
            store.writeJSON("catalog-cache.json", encode(data))
            return data
        }
        // Use disk or bundled titles so Home still renders.
        return readDisk() ?: fallbackBundle()
    }

    suspend fun search(query: String): List<MediaItem> {
        val needle = query.trim()
        if (needle.length < 2) return emptyList()
        val encoded = Json.formEncode(needle)
        val fetched: List<List<MediaItem>> = runParallel(
            listOf(
                { fetchCatalog("/catalog/movie/top/search=$encoded.json", "movie") },
                { fetchCatalog("/catalog/series/top/search=$encoded.json", "series") },
            ),
            emptyList(),
        )
        val live = dedupe(fetched[0] + fetched[1])
        if (live.isNotEmpty()) return live
        val lower = needle.lowercase(Locale.US)
        return fallback.filter {
            it.title.lowercase(Locale.US).contains(lower) ||
                (it.showTitle ?: "").lowercase(Locale.US).contains(lower)
        }
    }

    suspend fun genre(kind: String, name: String): List<MediaItem> {
        val key = "$kind:$name"
        synchronized(lock) {
            val hit = genreMem[key]
            if (hit != null && System.currentTimeMillis() - hit.first < 20 * 60 * 1000L) return hit.second
        }
        val encoded = queryEncode(name)
        val items = fetchCatalog("/catalog/$kind/top/genre=$encoded.json", kind)
        synchronized(lock) {
            genreMem[key] = System.currentTimeMillis() to items
        }
        return items
    }

    suspend fun meta(id: String): TitleMeta? {
        val imdb = TvmTitle.extractImdb(id)
            ?: return fallback.firstOrNull { it.id == id }?.let { TitleMeta(it, emptyList()) }
        synchronized(lock) {
            val hit = metaMem[imdb]
            if (hit != null && System.currentTimeMillis() - hit.first < 30 * 60 * 1000L) return hit.second
        }
        val fetched: List<TitleMeta?> = runParallel(
            listOf(
                { readKind("movie", imdb) },
                { readKind("series", imdb) },
            ),
            null,
        )
        // A title that exists as both prefers the series, which carries episodes.
        val picked = fetched[1] ?: fetched[0]
        if (picked != null) {
            synchronized(lock) {
                metaMem[imdb] = System.currentTimeMillis() to picked
            }
        }
        return picked
    }

    private fun readKind(kind: String, imdb: String): TitleMeta? {
        val body = getObject("https://v3-cinemeta.strem.io/meta/$kind/$imdb.json", SESSION_TIMEOUT_MS) ?: return null
        val raw = body.opt("meta") as? JSONObject ?: return null
        val item = mapMeta(raw, kind) ?: return null
        val children = if (kind == "series") mapVideos(item, raw.opt("videos")) else emptyList()
        return TitleMeta(item, children)
    }

    private fun fetchCatalog(path: String, kind: String): List<MediaItem> {
        val body = getObject("https://v3-cinemeta.strem.io$path", CATALOG_TIMEOUT_MS) ?: return emptyList()
        val metas = objectList(body.opt("metas")) ?: return emptyList()
        return metas.mapNotNull { mapMeta(it, kind) }
    }

    private fun mapMeta(raw: JSONObject, kind: String): MediaItem? {
        val id = (raw.opt("id") ?: raw.opt("imdb_id") ?: "").toString()
        if (!IMDB_ID.containsMatchIn(id)) return null
        val title = trimSpaces((raw.opt("name") as? String) ?: (raw.opt("title") as? String) ?: "")
        if (title.isEmpty()) return null
        val genres = stringList(raw.opt("genre")) ?: emptyList()
        val poster = (raw.opt("poster") as? String) ?: ""
        val backdrop = (raw.opt("background") as? String) ?: (raw.opt("backdrop") as? String) ?: poster
        val runtimeRaw = raw.opt("runtime")
        val runtime: String? = when {
            runtimeRaw is String -> runtimeRaw
            runtimeRaw is Int -> "$runtimeRaw min"
            runtimeRaw is Long -> "$runtimeRaw min"
            // NSNumber bridges to Int only when the value is whole; mirror that.
            runtimeRaw is Number && runtimeRaw.toDouble() == Math.floor(runtimeRaw.toDouble()) ->
                "${runtimeRaw.toLong()} min"
            else -> null
        }
        val ratingRaw = raw.opt("imdbRating")
        return MediaItem(
            id = id.lowercase(Locale.US),
            title = title,
            year = parseYear(raw.opt("year")) ?: parseYear(raw.opt("releaseInfo")),
            kind = kind,
            synopsis = (raw.opt("description") as? String) ?: "",
            poster = poster,
            backdrop = backdrop,
            genres = genres,
            rating = if (ratingRaw == null || ratingRaw === JSONObject.NULL) "" else ratingRaw.toString(),
            runtime = runtime,
            playable = true,
            progress = null,
            filename = null,
            hue = TvmTitle.hue(title),
            mimeType = null,
            season = null,
            episode = null,
            episodeName = null,
            showTitle = title,
            aired = null,
            added = null,
        )
    }

    private fun mapVideos(show: MediaItem, videos: Any?): List<MediaItem> {
        val list = objectList(videos) ?: return emptyList()
        val mapped = list.mapNotNull { video ->
            val fromId = (video.opt("id") ?: "").toString()
            val idParts = fromId.split(":").filter { it.isNotEmpty() }
            val season = Json.int(video.opt("season") ?: video.opt("seasonNumber"))
                ?: if (idParts.size >= 3) idParts[idParts.size - 2].toIntOrNull() else null
            val episode = Json.int(video.opt("episode") ?: video.opt("number") ?: video.opt("episodeNumber"))
                ?: idParts.lastOrNull()?.toIntOrNull()
            if (season == null || episode == null || season < 1 || episode < 1) return@mapNotNull null
            val name = (video.opt("title") as? String) ?: (video.opt("name") as? String) ?: "Episode $episode"
            val thumbnail = (video.opt("thumbnail") as? String)?.ifEmpty { null } ?: show.poster
            MediaItem(
                id = "${show.id}:$season:$episode",
                title = show.title,
                year = show.year,
                kind = "series",
                synopsis = (video.opt("overview") as? String) ?: (video.opt("description") as? String) ?: "",
                poster = thumbnail,
                backdrop = show.backdrop,
                genres = show.genres,
                rating = "",
                runtime = null,
                playable = true,
                progress = null,
                filename = null,
                hue = show.hue,
                mimeType = null,
                season = season,
                episode = episode,
                episodeName = name,
                showTitle = show.title,
                aired = null,
                added = null,
            )
        }
        return mapped.sortedWith(compareBy({ it.season ?: 0 }, { it.episode ?: 0 }))
    }

    private fun parseYear(value: Any?): Int? {
        val number = Json.int(value)
        if (number != null && number > 1800) return number
        val text = value as? String ?: return null
        val match = YEAR.find(text) ?: return null
        return match.value.toIntOrNull()
    }

    private fun dedupe(items: List<MediaItem>): List<MediaItem> {
        val seen = mutableSetOf<String>()
        return items.filter { seen.add(it.id) }
    }

    private fun fallbackBundle(): CatalogBundle {
        val movies = fallback.filter { it.kind == "movie" }
        val series = fallback.filter { it.kind == "series" }
        return CatalogBundle(
            moviesTop = movies.take(16),
            seriesTop = series.take(16),
            moviesRated = movies.take(16),
            seriesRated = series.take(16),
            recentTv = series.take(16),
            newFilms = movies.filter { (it.year ?: 0) >= 2023 }.take(16),
            catalog = fallback,
        )
    }

    private fun readDisk(): CatalogBundle? {
        val root = store.readJSON("catalog-cache.json") as? JSONObject ?: return null
        val movies = root.opt("moviesTop") as? JSONArray ?: return null
        return CatalogBundle(
            moviesTop = parseItems(movies),
            seriesTop = parseItems(root.opt("seriesTop")),
            moviesRated = parseItems(root.opt("moviesRated")),
            seriesRated = parseItems(root.opt("seriesRated")),
            recentTv = parseItems(root.opt("recentTv")),
            newFilms = parseItems(root.opt("newFilms")),
            catalog = parseItems(root.opt("catalog")),
        )
    }

    private fun encode(data: CatalogBundle): JSONObject = Json.obj(
        "moviesTop" to Json.array(data.moviesTop.map { it.json() }),
        "seriesTop" to Json.array(data.seriesTop.map { it.json() }),
        "moviesRated" to Json.array(data.moviesRated.map { it.json() }),
        "seriesRated" to Json.array(data.seriesRated.map { it.json() }),
        "recentTv" to Json.array(data.recentTv.map { it.json() }),
        "newFilms" to Json.array(data.newFilms.map { it.json() }),
        "catalog" to Json.array(data.catalog.map { it.json() }),
    )

    private fun parseItems(value: Any?): List<MediaItem> {
        val array = value as? JSONArray ?: return emptyList()
        return (0 until array.length()).mapNotNull { MediaItem.parse(array.opt(it)) }
    }

    /**
     * Swift's `as? [[String: Any]]` is all-or-nothing: one non-object entry
     * discards the whole list. Keep that, or a stray element would let entries
     * through that iOS drops.
     */
    private fun objectList(value: Any?): List<JSONObject>? {
        val array = value as? JSONArray ?: return null
        val out = ArrayList<JSONObject>(array.length())
        for (index in 0 until array.length()) {
            val entry = array.opt(index) as? JSONObject ?: return null
            out.add(entry)
        }
        return out
    }

    /** Same all-or-nothing rule as above, and unlike Json.strings it keeps empty entries. */
    private fun stringList(value: Any?): List<String>? {
        val array = value as? JSONArray ?: return null
        val out = ArrayList<String>(array.length())
        for (index in 0 until array.length()) {
            val entry = array.opt(index) as? String ?: return null
            out.add(entry)
        }
        return out
    }

    /** Foundation's `.whitespaces` is spaces and tabs only, so newlines survive the trim. */
    private fun trimSpaces(value: String): String = value.trim {
        it == '\t' || Character.getType(it) == Character.SPACE_SEPARATOR.toInt()
    }

    /**
     * Matches Foundation's `.urlQueryAllowed`, which leaves sub-delimiters
     * intact: a genre such as "Action & Adventure" must reach Cinemeta with its
     * ampersand literal, exactly as iOS sends it.
     */
    private fun queryEncode(value: String): String {
        val allowed = "-._~!$&'()*+,/:;=?@"
        val out = StringBuilder()
        for (byte in value.toByteArray(Charsets.UTF_8)) {
            val code = byte.toInt() and 0xFF
            val ch = code.toChar()
            if (code < 128 && (ch.isLetterOrDigit() || allowed.indexOf(ch) >= 0)) {
                out.append(ch)
            } else {
                out.append('%').append(String.format(Locale.US, "%02X", code))
            }
        }
        return out.toString()
    }

    private fun getObject(url: String, timeoutMs: Int): JSONObject? {
        var conn: HttpURLConnection? = null
        return try {
            val opened = URI(url).toURL().openConnection() as HttpURLConnection
            conn = opened
            opened.requestMethod = "GET"
            opened.connectTimeout = timeoutMs
            opened.readTimeout = timeoutMs
            opened.useCaches = false
            val code = opened.responseCode
            if (code !in 200..299) return null
            val text = opened.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() }
            JSONObject(text)
        } catch (error: Exception) {
            null
        } finally {
            conn?.disconnect()
        }
    }

    /**
     * The Swift fires these with `async let`. Without kotlinx-coroutines on the
     * classpath, plain daemon threads keep the same shape: six twelve-second
     * timeouts must not serialise behind one another.
     */
    private fun <T> runParallel(tasks: List<() -> T>, fallbackValue: T): List<T> {
        if (tasks.isEmpty()) return emptyList()
        val pool = Executors.newFixedThreadPool(tasks.size) { runnable ->
            Thread(runnable, "tvm-catalog").apply { isDaemon = true }
        }
        return try {
            val futures = tasks.map { task -> pool.submit(Callable<T> { task() }) }
            futures.map { future -> runCatching { future.get() }.getOrDefault(fallbackValue) }
        } finally {
            pool.shutdown()
        }
    }

    private companion object {
        const val CATALOG_TIMEOUT_MS = 12_000
        const val SESSION_TIMEOUT_MS = 20_000
        val IMDB_ID = Regex("""^tt\d+$""")
        val YEAR = Regex("""\b(19|20)\d{2}\b""")
    }
}
